### ISSUE TITLE
Jenkins pipeline and protractor

### ISSUE THREAD

**Issue by cbornet** [jhipster/generator-jhipster#4693]
This is to continue the discussion started in #4600 about Jenkins and protractor.

I tested https://github.com/mark-adams/docker-chromium-xvfb/tree/master/images/base to run chrome in docker and it works well.
I guess we can adapt to run firefox instead since that's what we set in the protractor conf file.
Now the difficulty with protractor is that it needs to start the application and you probably need other elements to run (Database, Elastic Search, JHipster Registry, etc...)
So I see several possibilities:
1. Don't run the build in a docker and ask the user to install the elements by himself. Not very nice...
2. Run the elements from Jenkins in docker containers and do the necessary port/host mapping (I don't know if it's easy without docker-compose)
3. Run the elements from the jhipster-ci-stack container (so you could have a dockerized Jenkins that launches a jhipster-ci-stack docker that launches a mysql docker. Looks like too much docker in docker in docker, but maybe that can work...). For comparison, Travis uses the docker-compose file in src/main/docker to launch the elements.
4. Put all the possible combinations of docker images on the docker hub (eg. jhipster-ci-stack-psql-es). Probably too much...
5. Eventually, we could generate a Dockerfile that is used by the Jenkinsfile to build himself the docker build container (see https://jenkins.io/blog/2016/08/08/docker-pipeline-environments/). I find this quite elegant.

Opinion ?

cc @mraible

**Comment by cbornet** [comment]
On second thought 2 and 3 would cause issues when multiple builds run in parallel.

**Comment by cbornet** [comment]
So it appears the best solution I found is actually 2 using docker container linking to give access to the containers. This is kinda link what is done in src/main/docker/app.yml but without docker-compose.
You can try the following pipeline script for instance
```groovy
node {
    stage('checkout') {
        git url: 'https://github.com/jhipster/jhipster-sample-app.git'
    }
    
    def mysql = docker.image('mysql:5.7.13').run('-e MYSQL_USER=root -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -e MYSQL_DATABASE=mysql')

    docker.image('openjdk:8').inside("--link ${mysql.id}:mysql -u root -e MAVEN_OPTS=\"-Duser.home=./\"") {
        stage('check tools') {
            sh "java -version"
        }
        
        stage('clean') {
            sh "./mvnw clean"
        }
        
        stage('run') {
            sh "echo '{ \"allow_root\": true }' > /root/.bowerrc"
            sh "export SPRING_DATASOURCE_URL=\"jdbc:mysql://mysql:3306/mysql?useUnicode=true&characterEncoding=utf8&useSSL=false\" && ./mvnw -Pprod"
        }
    }
```
Note: this script is not usable as-is as it will block at the end. A full script would need to run the server in background, wait that it is started and launch the protractor tests.

As this uses docker, it is dependent on #4600 being merged.
Also as there is quite a bit of work, are there people interested in having Protractor tests running in Jenkins (please thumb up if interested) ?

**Comment by jdubois** [comment]
Closing this as #4903 is merged

**Comment by cbornet** [comment]
Yes even though running the protractor tests is not part of the PR. But this will be a lot of work so no need to keep this issue opened.
